go/chat/wallet/parser.go

```diff
@@ -0,0 +1,112 @@
+package wallet
+
+import (
+	"regexp"
+	"strings"
+
+	chat1 "github.com/keybase/client/go/protocol/chat1"
+)
+
+var txPattern = regexp.MustCompile(
+	// Must explicitly have a + in front
+	`\+` +
+		// Stellar decimal amount
+		`(\d+\.?\d*|\d*\.?\d+)` +
+		// Currency code
+		`([A-Za-z]{2,6})` +
+		// At sign and username, optional for direct messages
+		// If not used, must be followed by a non-tx-character or end of string
+		`(?:(?:@((?:[a-zA-Z0-9]+_?)+))|(?:[^A-Za-z@]|\z))`,
+)
+
+var maxAmountLength = 100
+var maxUsernameLength = 16
+var maxTxsPerMessage = -1
+
+func findChatTxCandidates(xs string) []chat1.ChatTxCandidate {
+	// A string that does not appear in the candidate regex so we don't get false positives from concatenations.
+	replacer := "$"
+	replaced := replaceQuotedSubstrings(xs, replacer)
+
+	rawMatches := txPattern.FindAllStringSubmatch(replaced, maxTxsPerMessage)
+	matches := make([]chat1.ChatTxCandidate, 0, len(rawMatches))
+	for _, rawMatch := range rawMatches {
+		amount := rawMatch[1]
+		currencyCode := rawMatch[2]
+		username := rawMatch[3]
+		if len(amount) <= maxAmountLength && len(username) <= maxUsernameLength {
+			var txUsername *string
+			if username == "" {
+				txUsername = nil
+			} else {
+				txUsername = &username
+			}
+			matches = append(matches, chat1.ChatTxCandidate{Amount: amount, CurrencyCode: currencyCode, Username: txUsername})
+		}
+	}
+	return matches
+}
+
+var startQuote = ">"
+var startCodeLine = []rune("`")
+var startCodeBlock = []rune("```")
+var endCodeLine = []rune("`")
+var endCodeBlock = []rune("```")
+var newline = []rune("\n")
+
+type newlineBehavior = int
+
+const (
+	sameLine newlineBehavior = iota
+	multiLine
+)
+
+func replaceQuotedSubstrings(xs string, replacer string) string {
```

Comment:
Seems like you can just do what I did for unfurling here: https://github.com/keybase/client/blob/7a816baeb2b6616a23f7d3e8345807dc5cafd021/go/chat/unfurl/extractor.go#L62

Reply:
huh, that makes a lot of sense. For some reason I thought this wouldn't work but since there's no arbitrary nesting I guess regex is fine. One comment: this incorrectly parses

```
```hello`world``` 
```
to
```
world`
```
instead of empty string. What do you think about
```
    xs = regexp.MustCompile("(?s)```.*?```").ReplaceAllString(xs, replacer)
    xs = regexp.MustCompile("(?s)`.*?`").ReplaceAllString(xs, replacer)
```?

Reply:
Sure sounds good.